src/main/ruby/truffleruby/core/posix.rb

```diff
@@ -425,11 +425,50 @@ def self.read_string_native(io, length)
     end
   end
 
-  def self.read_to_buffer_native(io, length)
+  def self.execute_posix_read(io, buffer, length)
     fd = io.fileno
+    return Truffle::POSIX.read(fd, buffer, length) unless io.timeout
+
+    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + Truffle::KernelOperations.convert_duration_to_seconds(io.timeout)
+
+    while true
+      bytes_read = Truffle::POSIX.read(fd, buffer, length)
+      return bytes_read if bytes_read >= 0
+      return bytes_read unless Errno.errno == Errno::EAGAIN::Errno
+
+      current_timeout = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      raise IO::TimeoutError if current_timeout <= 0
+
+      poll_result = Truffle::IOOperations.poll(io, Truffle::IOOperations::POLLIN, current_timeout)
+      raise IO::TimeoutError if poll_result == 0
+      Errno.handle_errno(Errno.errno) if poll_result == -1
+    end
+  end
+
+  def self.execute_posix_write(io, buffer, length)
+    fd = io.fileno
+    return Truffle::POSIX.write(fd, buffer, length) unless io.timeout
+
+    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + Truffle::KernelOperations.convert_duration_to_seconds(io.timeout)
```

Comment:
Wondering if we need more precise current time, e.g. in milliseconds or even microseconds, similar to other helper methods in the `IOOperations` module:
- `Process.clock_gettime(Process::CLOCK_MONOTONIC, :microsecond)`
- `Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)`

Reply:
So looking at both our implementation and the used posix poll - they both use milliseconds. However we both expect a float second value, as well we would need to convert the `Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)` from int to float - so rather than doing multiple conversions I just send the default seconds float output of `Process.clock_gettime(Process::CLOCK_MONOTONIC)` which is at least millisecond precise, and let the poll truncate what it doesn't need.
It's more about not doing redundant conversions than speed, though there is an insignificant speed benefit of this. Let me know if you feel otherwise.

Reply:
The logic above uses `Float` that's pretty slow and should be avoided, better to use nanos or millis.
In general nanos is the most general, so probably we should use that consistently.

Reply:
I see. I presume you're talking about the overhead after `Truffle::KernelOperations.convert_duration_to_*`. Currently `Truffle::IOOperations.poll` is also fixed to work with a float. Are you suggesting to refactor that method too?

Reply:
Re `Truffle::IOOperations.poll`, indeed, although so far it was only called fairly directly from user-facing methods, and not in a loop.
I think it would be extract the seconds -> milliseconds coercion part of poll(), and have something like `poll_optional_timeout` for existing usages and `poll_milliseconds_timeout` for this new usage.

Reply:
> I presume you're talking about the overhead after `Truffle::KernelOperations.convert_duration_to_*`.

Not so much because that's only called once. It's more like conversions between Float and Integer on every loop iteration below.

Reply:
Updated.